Generic tie-down/haul system for furniture on vehicles

RenechCDDA opened the pull request:
#### Summary
Features "Generic tie-down/haul system for furniture on vehicles"

#### Purpose of change
![image](https://github.com/user-attachments/assets/3dc4217f-648c-4012-af50-569a357ff5d8)


#### Describe the solution
Basically we make a new part with a bespoke action that only accepts furniture, some safety measures to make sure it can only pick up/place furniture in reasonable ways, and we store the furniture's raw string as an item var.

The storing part is what makes it all work.

We also increase the weight of the vehicle via the a new furniture field `mass` if any furniture is being stored. Currently the 'mass' of every furniture is simply 1000kg. The extra weight is removed when the stored furniture is removed from the vehicle part.

We append the name of the vehicle part using a bad hack stolen from the animal compartment. (This whole thing is largely a copy of the animal compartment's method)

#### Describe alternatives you've considered
Instead of storing it as a raw string in an item var we could store it as a furn_str_id pointer(?) and carry it with the vehicle part. That would possibly allow for easier future implementation of migration... but we have no such migration coverage.

#### Testing
Install or find any empty "cargo space". Pick up a bookcase. Put down a bookcase. Pick up, put down, repeat a few more times to make sure it works both ways. Pick up once more, try to put it down on the vehicle, nothing happens but information/failure message is added to log. 

![image](https://github.com/user-attachments/assets/80eca236-75b0-4c32-a38d-c239c7e16ac4)


#### Additional context

KNOWN ISSUES:
~~No migration support for stored furniture~~ --> some sort of migration support for them, based on whatever Procyonae made

You can tie down a furniture, uninstall the part, and then magically carry the furniture around with you (because it's stored as an item var on the base part). The weight increase does not apply to the item form, only as a vehicle part.

Furnitures do not have "weight" so this adds a flat **1000kg** to the vehicle whenever any furniture is being stored. Maybe we could make it a function of the furniture's movecost instead, but I'm not sure that is any better.

No sort of sprite support or anything.

Existing tie-down furnitures are still there, somebody needs to obsolete them (and maybe we need to migrate them or something?)

There is no time/activity taken to load/unload the furniture, it happens instantly

RenechCDDA committed:
Generic tie-down/haul system for furniture on vehicles

Procyonae commented:
My main concern with this is it makes future expansion of furniture more awkward bc this is only storing the id not a pointer (although I appreciate that's much more awkward to do bc you suddenly need to be able to serialise and deal with furnitures that aren't physically on the map)

> Furnitures do not have "weight" so this adds a flat 1000kg to the vehicle whenever any furniture is being stored. Maybe we could make it a function of the furniture's movecost instead, but I'm not sure that is any better.

We should just actually add furniture weights so I wouldn't worry about that for now

> Existing tie-down furnitures are still there, somebody needs to obsolete them (and maybe we need to migrate them or something?)

I can probably handle migrating them in a followup PR

Procyonae commented on src/vehicle_use.cpp:
```diff
@@ -1582,6 +1582,97 @@ void vehicle::use_monster_capture( int part, map */*here*/, const tripoint_bub_m
+        const std::function<bool( const tripoint_bub_ms & )> tile_without_furniture_obstruction =
+        [&here]( const tripoint_bub_ms & target ) {
+            // target does not have furniture or any vehicle part
+            // TODO: Where else can't we place a new furniture? There must be a common function for this
```
Probably want to modify game::can_move_furniture and use that

ehughsbaird commented on src/vehicle_use.cpp:
```diff
@@ -1582,6 +1582,97 @@ void vehicle::use_monster_capture( int part, map */*here*/, const tripoint_bub_m
+    // This is used to highlight only squares with moveable furniture
+    const std::function<bool( const tripoint_bub_ms & )> tile_with_furniture =
+    [&here]( const tripoint_bub_ms & target ) {
+        if( here->has_furn( target ) && here->furn( target )->move_str_req >= 0 ) {
```
You can give clearer feedback if you let them select any adjacent tile, then back out if the tile they pick doesn't have furniture or has furniture that can't be moved.

Also consider checking the avatar's strength for this.

RenechCDDA replied:
I think choose_adjacent_highlight() is much better when dealing with few conditions, but if I do add a strength check that'll be 3 separate conditions - at that point it starts to warrant more specific messaging, as you say. So let me see about this.

ehughsbaird commented on src/vehicle_use.cpp:
```diff
@@ -1582,6 +1582,97 @@ void vehicle::use_monster_capture( int part, map */*here*/, const tripoint_bub_m
+        furn_str_id tied_down_furniture( base_copy.get_var( item_var_string ) );
+        if( !tied_down_furniture.is_valid() ) {
+            debugmsg( "Invalid stored furniture %s", base_copy.get_var( item_var_string ) );
+            base_copy.erase_var( item_var_string ); // to prevent bricking the item
```
Because you're operating on a copy, you don't actually remove it from the vehicle part.

RenechCDDA replied:
We set our copy as the base part later on via ` parts[part].set_base( std::move( base_copy ) )` which handles it.  This is actually cargo culted from the use_monster_capture() function right above it. I'm very much unsure of why it does it that way (possibly because they call an iuse?), hence the comment. 

I was imagining there might be some reason not to modify the base part directly, but I did not investigate very far.

(Just wanted to reply to this one for the moment - gotta go through the rest later)

ehughsbaird replied:
Except in this case, where you return from the function without doing that. Sorry, my comment doesn't cover the return statement.

RenechCDDA replied:
...Right! Good catch.

ehughsbaird commented on src/vehicle_use.cpp:
```diff
@@ -1582,6 +1582,97 @@ void vehicle::use_monster_capture( int part, map */*here*/, const tripoint_bub_m
+
+    const std::string item_var_string = "tied_down_furniture";
+
+    // Why do we construct a copy here?
+    item base_copy = item( parts[part].get_base() );
```
Perhaps to not leave things in a broken intermediate state? It doesn't seem necessary here.

RenechCDDA replied:
get_base() returns a const reference, it turns out. Probably best to keep it keep it like that.

ehughsbaird reviewed (commented):
I think storing the id string is appropriate and I wouldn't worry about future changes in furniture.

I think it is also a good idea to handle removal/destruction of the part here before it is merged. For destruction, I recommend using the bash drops of the furniture.

Termineitor244 commented:
I think the same as @PatrikLundell, requiring actual installing of the new vehicle part as a normal cargo space (With a welder, rods, etc.) defeats much of the purpose of transporting furniture in your vehicle, which is mainly useful early on in the game, and it feels a little ridiculous to have to install a new cargo space (And the extra vehicle frame for it, or the removal of a seat or an actual cargo space already installed in the vehicle) just for the ability to load a fridge in the back of your pick up truck (Which is already possible with just a couple of ropes, as in reality).

I suggest just adding some different install requirements asking for ropes like the current tied-down vehicle parts, and limiting the installing of this new vehicle part to places where there is already a cargo space installed (Not sure if this part is actually possible though).

Procyonae commented:
> I think storing the id string is appropriate

It's appropriate as long as we don't have important metadata for furniture, currently we only have partial damage iirc but eg I have a branch adding paint metadata to ter+furn (which mostly works but I need to support adding it at mapgen etc) which would obviously be lost through this. I agree it shouldn't block this PR tho, just might require !!FUN!! migration in the future.

RenechCDDA commented on src/vehicle_part.cpp:
```diff
@@ -119,9 +120,18 @@ std::string vehicle_part::name( bool with_prefix ) const
+    // furniture tiedown
+    if( base.has_var( "tied_down_furniture" ) ) {
+        furn_str_id stored_furniture( base.get_var( "tied_down_furniture" ) );
+        if( !stored_furniture.is_valid() ) {
+            return res; // very bad, but furn_str_id constructor will error so we don't need debugmsg here
+        }
+        res += string_format( _( " holding %s" ), stored_furniture->name() );
```
You're just plain right here.

RenechCDDA committed:
Various improvements after code review from Ehughsbaird

-Add strength/lifting check when loading furniture, split it up for more specific feedback when something can't be loaded
-Switch the furniture's weight to being a member of the furniture class, loaded from its object (get rid of stored_furniture_weight)
-Don't error/return early on UI display when underlying carried furniture is invalid
-Add a feedback message when trying to unload/load furniture from a broken part
-Use furniture name where possible instead of generically referring to "the furniture"
-Make sure we actually clear the item var when panic aborting

RenechCDDA committed:
Break up vehicle::use_tiedown_furniture()

Procyonae commented on src/vehicle_part.cpp:
```diff
@@ -119,9 +120,21 @@ std::string vehicle_part::name( bool with_prefix ) const
+        if( stored_furniture.is_valid() ) {
+            res += string_format( _( " holding %s" ), stored_furniture->name() );
+        } else {
+            // no debugmsg or else it will trigger every frame, essentially forcing (i)gnore error to exit the menu at all
```
Shouldn't it debugmsg and then remove the furniture var so it doesn't error again when it's set and during deserialisation? You also need to handle furniture migration at deserialisation of the variable or when it's first accessed.

Procyonae replied:
> Item vars don't have any sort of handling for that.

Well no you'll need to add the handling to Item_factory::check_definitions()

Procyonae replied:
> Item_factory::check_definitions() iterates itypes

Oh right >.<

This compiles and loads a save without exploding, not tested it with your branch tho
```diff --git a/src/savegame_json.cpp b/src/savegame_json.cpp
index 739d7978b7..18b9861e59 100644
--- a/src/savegame_json.cpp
+++ b/src/savegame_json.cpp
@@ -178,6 +178,8 @@ static const ter_str_id ter_t_switchgear_l( "t_switchgear_l" );
 static const ter_str_id ter_t_switchgear_s( "t_switchgear_s" );
 static const ter_str_id ter_t_wreckage( "t_wreckage" );

+static const std::string TIED_DOWN_FURNITURE = "tied_down_furniture";
+
 static const std::array<std::string, static_cast<size_t>( object_type::NUM_OBJECT_TYPES )>
 obj_type_name = { { "OBJECT_NONE", "OBJECT_ITEM", "OBJECT_ACTOR", "OBJECT_PLAYER",
         "OBJECT_NPC", "OBJECT_MONSTER", "OBJECT_VEHICLE", "OBJECT_TRAP", "OBJECT_FIELD",
@@ -185,6 +187,9 @@ obj_type_name = { { "OBJECT_NONE", "OBJECT_ITEM", "OBJECT_ACTOR", "OBJECT_PLAYER
     }
 };

+static std::unordered_map<ter_str_id, std::pair<ter_str_id, furn_str_id>> ter_migrations;
+static std::unordered_map<furn_str_id, std::pair<ter_str_id, furn_str_id>> furn_migrations;
+
 // TODO: investigate serializing other members of the Creature class hierarchy
 static void serialize( const weak_ptr_fast<monster> &obj, JsonOut &jsout )
 {
@@ -2891,6 +2896,14 @@ void item::io( Archive &archive )
             }
         }
     }
+
+    if( auto var = item_vars.find( TIED_DOWN_FURNITURE ); var != item_vars.end() ) {
+        const furn_str_id furnstr( var->second.str() );
+        if( auto it = furn_migrations.find( furnstr ); it != furn_migrations.end() ) {
+            set_var( TIED_DOWN_FURNITURE, it->second.second.str() );
+        }
+    }
+
     // TODO: change default to empty string
     archive.io( "name", corpse_name, std::string() );
     archive.io( "owner", owner, faction_id::NULL_ID() );
@@ -4697,9 +4710,6 @@ void _write_rle_terrain( JsonOut &jsout, std::string_view ter, int num )

 } // namespace

-static std::unordered_map<ter_str_id, std::pair<ter_str_id, furn_str_id>> ter_migrations;
-static std::unordered_map<furn_str_id, std::pair<ter_str_id, furn_str_id>> furn_migrations;
-
 void ter_furn_migrations::load( const JsonObject &jo )
 {
     //TODO: Add support for migrating to items?

```

RenechCDDA replied:
Okay seems to work, so yolo commit it is

![image](https://github.com/user-attachments/assets/ae135dbf-294c-401f-9e48-e10e49b5ac6d)

db48x commented:
I love the idea, but I also agree that requiring welding to convert a `cargo_space` into a `cargo_furniture_tiedown` is odd. What if all `cargo_spaces` had furniture tiedowns (just mention them in the description)? Then, provide an action that takes ropes or seatbelts or whatever, ties them to the tiedowns, and otherwise makes it ready to accept furniture. That action can either clear away the loose items, or require that there be no loose items. Or perhaps when the player attempts to place furniture in it it could just bail early if loose items are present, printing a log message to tell the player why.

That would seem to kill two birds with but one stone.

RenechCDDA committed:
Allow either furniture or items, but not both

RenechCDDA commented:
Okay, fine. That turns out to be a more or less one-line change, just predicated on *exacting* knowledge of how class inheritance works and some prior efforts to spread item_stack throughout our codebase.


(Note: While testing this I discovered the AIM's **display** shows 0/500L for volume, but still correctly prevents you from shoving things into a space where there is a furniture.)

RenechCDDA committed:
Tied-down furniture migration, or something
